src/data_model/transformations.jl

```diff
@@ -472,3 +472,917 @@ function apply_voltage_angle_difference_bounds!(eng::Dict{String,<:Any}, vad::Re
+        for (_,comp) in data_eng[type]
+            if haskey(comp, "configuration")
+                conf = comp["configuration"]
+                if isa(conf, Vector)
+                    for w in 1:length(conf)
+                        if conf[w]==WYE
+                            push!(starpoints, (comp["bus"][w], comp["connections"][w][end]))
+                        end
+                    end
+                else
+                    if conf==WYE
+                        push!(starpoints, (comp["bus"], comp["connections"][end]))
+                    end
+                end
+            end
+        end
+    end
+
+    # find grounded terminals
+    grounded_terminals = [(b,t) for (b,bus) in data_eng["bus"] for t in bus["grounded"]]
+
+    # assign initial neutral terminals
+    nbts = [starpoints..., grounded_terminals...]
+
+    # propagate neutral terminals through galvanic connections
+    stack = deepcopy(nbts)
+    while !isempty(stack)
+        bt = pop!(stack)
+        new_nbts = setdiff(bt_neighbours[bt], nbts)
+        append!(stack, new_nbts)
+        append!(nbts, new_nbts)
+    end
+
+    return nbts
+end
+
+
+"Reduces line models, by removing trailing lines and merging lines in series with the same linecode."
+function reduce_lines!(data_eng::Dict{String,Any})
+    delete_trailing_lines!(data_eng)
+    join_lines!(data_eng)
+end
+
+
+"Returns a reduced data model, by removing trailing lines and merging lines in series with the same linecode."
+reduce_lines(data_eng::Dict{String,Any}) = reduce_lines!(deepcopy(data_eng))
+
+
+"Reverse the direction of a line."
+function _line_reverse!(line::Dict{String,Any})
+    prop_pairs = [("f_bus", "t_bus"), ("g_fr", "g_to"), ("b_fr","b_to")]
+
+    for (x,y) in prop_pairs
+        if haskey(line, x)
+            tmp = line[x]
+            line[x] = line[y]
+            line[y] = tmp
+        end
+    end
+end
+
+
+"""
+Returns a unique list of all buses specified in the data model.
+The argument 'comp_types' specifies which component types are searched to build the list.
+"""
+function get_defined_buses(data_eng::Dict{String,Any}; comp_types=pmd_eng_asset_types)
+    @assert data_eng["data_model"]==ENGINEERING
+
+    buses_exclude = []
+    for comp_type in intersect(comp_types, keys(data_eng))
+        for (id,comp) in data_eng[comp_type]
+            if haskey(comp, "f_bus")
+                buses = [comp["f_bus"], comp["t_bus"]]
+            elseif haskey(comp, "bus")
+                # works for a vector of buses and a single bus as a string
+                buses = isa(comp["bus"], String) ? [comp["bus"]] : comp["bus"]
+            else
+                buses = []
+            end
+            append!(buses_exclude, buses)
+        end
+    end
+    
+    return unique(buses_exclude)
+end
+
+
+"""
+Deletes trailing lines, 
+i.e. lines connected to a bus with no other connected components and which is not grounded.
+"""
```

Comment:
For exported functions, we should follow the new docstring pattern where the function is given (indented) first, see, e.g. transform_data_model function as an example.